pkg/discovery/discovery.go

```diff
@@ -151,10 +150,19 @@ func Run(restConf *rest.Config, cfg *config.Config) (errCount int) {
 		trackErrorsFor("querying resources under namespace " + ns)(
 			QueryResources(apiHelper, recorder, nsResources, &ns, cfg),
 		)
+	}
 
-		trackErrorsFor("querying pod logs under namespace " + ns)(
-			QueryPodLogs(kubeClient, recorder, ns, cfg),
-		)
+	// query pod logs
+	if cfg.Resources == nil || sliceContains(cfg.Resources, "podlogs") {
+		preprocessPodLogLimits(cfg, kubeClient)
+		nsListLogs, _ := FilterNamespaces(kubeClient, ".*")
+		for _, ns := range nsListLogs {
+			trackErrorsFor("querying pod logs under namespace " + ns)(
+				QueryPodLogs(kubeClient, recorder, ns, cfg),
```

Comment:
This flow is pretty unintuitive; I see what you're doing now that I've looked into the implementation of each function but it shouldn't require that.

I think it would be more clear (and testable) to do the following:

 - use the config to get a list of namespaces to query for logs
   - use that list of namespaces in the `for` loop
 - within the `QueryPodLogs` options, generate the necessary field selectors for the query. This is somewhat inefficient since you'll do it each time you query a namespace but the code will be more readable, we'll get rid of methods that mutate the config, and you'll have a few simple-to-test functions for getting the namespaces/field selectors from a static piece of data. 
   - in the existing implementation we also end up filtering the namespaces inside the preprocessing function and right afterwards; the extra API call is probably more expensive than iterating over our struct a few more times to make those field selectors.

Reply:
The thing is, it's a bit difficult to get the targeted namespaces, which could be defined in FieldSelectors. Then we need to parse the FieldSelectors ourselves, validating it, take out the selector which contains "metadata.namespace=xx", and append to the targeted namespace list. 

Comparing to all these additional effort, the current solution is will not take time more than O(N), and all invalidate namespace will be filtered out when doing list of pods query. We will only proceed to get pod logs for valid pods.

Reply:
looking into https://github.com/kubernetes/client-go/issues/410

Reply:
After discussion, here is what we are proposing:
1) get a list of namespaces by: Namespaces regex and SonobuoyNamespace bool
2) for each namespace, get all pod logs clientset.CoreV1().Pods(ns).List(...)
3) then do another query with field selectors, get pod logs clientset.CoreV1().Pods("").List(field selectors)

Then the OR logic will be fulfilled:
`(Namespaces OR SonobuoyNamespace OR FieldSelectors) `

Reply:
unit testing done.